crates/ty_python_semantic/src/types/tuple.rs

```diff
@@ -226,23 +219,22 @@ impl<'db> TupleType<'db> {
         TupleType::new(db, TupleSpec::homogeneous(element))
     }
 
+    // N.B. If this method is not Salsa-tracked, we take 10 minutes to check
+    // `static-frame` as part of a mypy_primer run! This is because it's called
+    // from `NominalInstanceType::class()`, which is a very hot method.
     #[salsa::tracked]
-    pub(crate) fn to_class_type(self, db: &'db dyn Db) -> Option<ClassType<'db>> {
-        KnownClass::Tuple
+    pub(crate) fn to_class_type(self, db: &'db dyn Db) -> ClassType<'db> {
+        let tuple_class = KnownClass::Tuple
             .try_to_class_literal(db)
-            .and_then(|class_literal| match class_literal.generic_context(db) {
-                None => Some(ClassType::NonGeneric(class_literal)),
-                Some(generic_context) if generic_context.variables(db).len() != 1 => None,
-                Some(generic_context) => Some(
-                    class_literal
-                        .apply_specialization(db, |_| generic_context.specialize_tuple(db, self)),
-                ),
-            })
-    }
+            .expect("Typeshed should always have a `tuple` class in `builtins.pyi`");
```

Comment:
This `.expect()` call means we now panic if we're given a custom typeshed without a `tuple` class. I don't _love_ this but I also think it's _okay_ -- there are already classes where we panic if they're not available in typeshed, such as `object` and `types.ModuleType`. `tuple` is similarly special; I think it's okay if we effectively mandate that custom typesheds have to have a `tuple` class in them.

Doing the refactor in the third commit of this PR (which provided a big performance boost) without the `.expect()` call here would be quite a bit harder, I think.